pilot/pkg/serviceregistry/kube/controller/endpointslice.go

```diff
@@ -59,6 +61,14 @@ func newEndpointSliceController(c *Controller) *endpointSliceController {
 	return out
 }
 
+// getEndpointKey generates a unique key for the endpoints based on ip addresses, portName and portNum.
+func getEndpointKey(portName string, portNum int32, ips []string) string {
```

Comment:
I'm guessing there might be a performance issue here. It might be better to use string builder overall.

Reply:
Hi @wulianglongrd, I understand that you mean I need to convert `portNum` into a string before using `fmt.Sprintf`?

Reply:
```
func BenchmarkGetEndpointKey(b *testing.B) {
	ips := []string{"127.0.0.1", "1.2.3.4"}

	b.Run("old", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			getEndpointKey("foo", 80, ips)
		}
	})
	b.Run("new", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			getEndpointKeyV2("foo", 80, ips)
		}
	})
}

// This is the current implementation
func getEndpointKey(portName string, portNum int32, ips []string) string {
	var ipaddrs []string
	ipaddrs = append(ipaddrs, ips...)
	ipString := strings.Join(ipaddrs, ", ") // why is there a blank needed here?
	return fmt.Sprintf("%s-%s-%d", ipString, portName, portNum)
}

// This implementation performs better.
func getEndpointKeyV2(portName string, portNum int32, ips []string) string {
	var b strings.Builder
	for k, ip := range ips {
		if k > 0 {
			b.WriteString(", ")
		}
		b.WriteString(ip)
	}

	b.WriteString("-")
	b.WriteString(portName)
	b.WriteString("-")
	b.WriteString(strconv.Itoa(int(portNum)))

	return b.String()
}
```
benchmark result:
```
$ go test -run="none" -bench="BenchmarkGetEndpointKey" -test.benchmem . -count=3
goos: darwin
goarch: amd64
pkg: github.com/wulianglongrd/learning-golang/demo
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkGetEndpointKey/old-12         	 3748388	       317.4 ns/op	     120 B/op	       5 allocs/op
BenchmarkGetEndpointKey/old-12         	 3886032	       310.5 ns/op	     120 B/op	       5 allocs/op
BenchmarkGetEndpointKey/old-12         	 3798856	       311.7 ns/op	     120 B/op	       5 allocs/op
BenchmarkGetEndpointKey/new-12         	15885232	        75.82 ns/op	      48 B/op	       2 allocs/op
BenchmarkGetEndpointKey/new-12         	15971180	        74.67 ns/op	      48 B/op	       2 allocs/op
BenchmarkGetEndpointKey/new-12         	16095555	        75.01 ns/op	      48 B/op	       2 allocs/op
PASS
ok  	github.com/wulianglongrd/learning-golang/demo	8.766s
```

Reply:
Hi @wulianglongrd, thanks for the performance test. I noticed that there are primary 2 changes:
1. convert int32 into string for `portNum`
2. use strings.Builder to join strings instead `fmt.Sprintf`
Which change do you think is the most important for this enhancement?

Reply:
`fmt.Sprintf` i think

Reply:
If so, I can replace `fmt.Sprintf` with `strings.Join`.

Reply:
To clarify, I mean the `fmt.Sprintf` function has low performance. It is about 50% lower than concat using `+` (see benchmark #48683 ). However, this is a big PR, these optimizations are probably trivial, and I'm a bit nitpicky and ignore it LGTM.

Reply:
Hi @wulianglongrd, it's okay, I prefer `strings.Join` to concat all string together, please check my newest submission, and thank you so much for your suggestion! ^_^